### ISSUE TITLE
Kubernetes version dropdown is empty

### ISSUE THREAD

**Issue by loganhz** [rancher/rancher#24091]
Version:

2.3.2

Steps:

1. Start rancher server
2. Create a RKE cluster

Results:

Kubernetes version dropdown is empty

![image](https://user-images.githubusercontent.com/21168270/68823064-a08ed900-06cd-11ea-97a9-66e449451f39.png)

Expected:

Users should be able to see k8s versions.

More Info:

If the connection between rancher server and github is bad, rancher server might take 30 mins to finish the `git clone k8s-metadata`. During this 30 mins, users will not be able to see any k8s version in create cluster page.

**Comment by shikanon** [comment]
yes, I will meet this same problem

**Comment by zhkou** [comment]
+1  
![image](https://user-images.githubusercontent.com/45305818/69821526-e3f55580-123e-11ea-9037-8da014facebd.png)

**Comment by dan-evolvere** [comment]
cant see this mentioned already, but this is happening in a fresh 2.3.3 aswell

**Comment by abhi1693** [comment]
I have this issue on a fresh v2.3.3 as well

**Comment by abhi1693** [comment]
It worked for me after I refreshed the Kubernetes metadata from `Tools -> Drivers -> Refresh Kubernetes Metadata`

**Comment by dan-evolvere** [comment]
Confirmed, that also fixed it for me.

**Comment by jiaqiluo** [comment]
the bug fix is validated in Rnacher:v2.3-head ab7ce7ba2 and master-head 8bd9f3be2

Steps:
- run a new Rancher server container
- log into the UI for the first time
- check if the dropdown list is empty
- check Rancher's log
- repeat the above steps several times 

Results:
- the dropdown list os not empty in all my tests 
- the log shows the following message, which indicates the fix is working, otherwise the dropdown list may be empty. 
```
2020/01/03 00:53:02 [INFO] driverMetadata: empty setting k8s-version, using stored defaults
```

**Comment by sowmyav27** [comment]
**Issue seen on 2.3.4rc6. Reopening the issue**

**Steps:**
- Have a rancher server set up in HA mode on 2.3.3 (steps followed from rancher docs, using helm 2)
- create a cluster, rke DO, the k8s version dropdown will be seen empty.
- refresh the kontainer metada from Global --> Tools --> Drivers
- Create an rke cluster, the k8s version will be available in the dropdown. Deploy a cluster.
- Upgrade rancher to 2.3.4-rc6
- Go to edit cluster and user can see k8s version dropdown will be seen empty.
- refresh the kontainer metada from Global --> Tools --> Drivers
- Create an rke cluster, the k8s version will be available in the dropdown.

**Comment by kinarashah** [comment]
@sowmyav27 After upgrade, do you see `"driverMetadata: empty setting k8s-version, using stored defaults"` in the logs? An upgraded setup would already have some set values, so I wonder what's happening here

**Comment by sowmyav27** [comment]
@kinarashah I see this in the logs
```
2020/01/03 05:33:23 [INFO] driverMetadata: refresh data
2020/01/03 05:33:25 [INFO] driverMetadata initialized successfully
```

**Comment by maggieliu** [comment]
Moving out of 2.3.4, need more time to investigate and focus on the zero downtime.

**Comment by r9labs** [comment]
Just a heads up, for those offline deployments, this workaround doesn't work, and it is still a problem.

**Comment by wzrdtales** [comment]
Had the same problem today. The problem is that the information of available versions is no longer coupled with the rancher bin, but seems to be retrieved from github now. However, the environment this was installed was locked down and github was blocked. After unblocking github and using the refresh metadata button mentioned here earlier it started working.

@superseb This might be just a task of documentation and not necessarily a bug fix I think. It would have helped us tremendously to know that github is now a requirement for rancher to function properly. And it would be a great help if you could document all the urls you're trying to reach from rancher to guarantee its function. A lot of companies want to lock down github in fear of too easy introduction of backdoors (I know it is not quite reasonable when at the same time docker hub and others are freely open though).

**Comment by dan-evolvere** [comment]
while what @wzrdtales says about github access may ring true in orgs that allow github/dockerhub, in a properly airgapped environment, this will never be allowed. Given that we already have a process to mirror images into a private registry for curation etc, would it not make sense to have a similar capability for k8s versions? If Rancher is presented as a potential self service offering for users, having air gapped control over this kind of setting/config as an admin makes a lot of sense to me.

**Comment by kinarashah** [comment]
A couple of things, 
- This feature was a part of release 2.3, which allows introducing new k8s versions dynamically to a rancher server setup. While this could have been highlighted more (by mentioning it under Breaking changes or the like), it was mentioned under Features and Enhancements, and also points to the doc link. https://rancher.com/docs/rancher/v2.x/en/admin-settings/k8s-metadata/ 

- For airgap setups or when the url fetch fails, rancher falls back to the versions coupled with the rancher bin as you pointed out. It's a bug if it doesn't. The original issue was happening in setups with very low internet speed where there was no error, it was just taking longer than expected.

**Comment by divyangjp** [comment]
Can confirm that it still is happening with `v2.3.5`. Metadata refresh works only if the user has `admin` privileges. Standard users face error when doing metadata refresh.

**Comment by DJAyth** [comment]
Running into this as well on new v2.3.5 installs, however, I'm trying to use terraform so a manual step of refreshing the metadata is not a good option. Any ideas?

**Comment by kinarashah** [comment]
Should be fixed in 2.4, improvements around loading metadata on startup by 
https://github.com/rancher/rancher/issues/25162

**Comment by kinarashah** [comment]
@divyangjp correct, the option to refresh the Kubernetes metadata is available for administrators by default, or for any user who has the Manage Cluster Drivers global role

**Comment by jiaqiluo** [comment]
the fix is validated in rancher master-head 854e9ae6f. 

Following the same steps provided in the issue, I do not see the dropdown list empty, and Rancher logs show the following message:
```
2020/02/28 23:48:09 [INFO] Refreshing driverMetadata in 1440 minutes 
2020/02/28 23:48:09 [INFO] driverMetadata: refreshing data from upstream https://releases.rancher.com/kontainer-driver-metadata/dev-v2.4/data.json
```

The same test is done multiple times and the issue is not seen.
